src/compiler/crystal/semantic/restrictions.cr

```diff
@@ -181,9 +192,71 @@ module Crystal
         return false
       end
 
+      # Check free vars (using `index` to generate unique vars for each underscore)
+      self_free_vars = all_free_vars.map_with_index do |(free_var, _), index|
+        free_var == "_" ? index : free_var
+      end
+      other_free_vars = all_free_vars.map_with_index do |(_, free_var), index|
+        free_var == "_" ? index : free_var
+      end
+
+      # If the other uses the same free var for different free vars in self,
+      # then self is never stricter than other (e.g. `T, _` vs `T, T`)
+      (0...all_free_vars.size).group_by { |index| other_free_vars[index] }.each_value do |indices|
+        opposite_free_vars = indices.map { |i| self_free_vars[i] }
+        return false if opposite_free_vars.any? { |var| var != opposite_free_vars.first }
+      end
+      # If self uses the same free var for different free vars in the other,
+      # then self is stricter
+      (0...all_free_vars.size).group_by { |index| self_free_vars[index] }.each_value do |indices|
+        opposite_free_vars = indices.map { |i| other_free_vars[i] }
+        return true if opposite_free_vars.any? { |var| var != opposite_free_vars.first }
+      end
+
       true
```

Comment:
What's the point of `return true` in the iteration when that's the only remaining return value of the method anyways? There don't seem to be any side effects, so the iteration would be unnecessary.

Reply:
That's in case more checks are added and someone forgets to do the check in this direction. Do I simply comment out this section?

Reply:
This explanation as a code comment would be great 👍